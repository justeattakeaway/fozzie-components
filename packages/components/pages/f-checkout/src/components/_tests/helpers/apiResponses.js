import getCheckoutResponse from '../../../../stories/api/responses/getCheckout/getCheckout';
import getBasketResponse from '../../../../stories/api/responses/getBasket/getBasket';
import { getAvailableTimes } from '../../../../stories/api/responses/getAvailableTimes';
import geoLocationDetailsResponse from '../../../../stories/api/responses/getGeoLocation';
import { getNotesConfig } from '../../../../stories/api/responses/getNotes';
import { CHECKOUT_METHOD_DELIVERY } from '../../../constants';

// getGeoLocation
export const geoLocationDetails = geoLocationDetailsResponse[0].payload;

// getCheckout
export const checkoutDelivery = getCheckoutResponse();

// getBasket
export const basketDelivery = getBasketResponse();
export const basketInvalidProducts = getBasketResponse(CHECKOUT_METHOD_DELIVERY, 'invalidProduct');
export const basketOfflineProducts = getBasketResponse(CHECKOUT_METHOD_DELIVERY, 'offlineProduct');
export const basketDeliveryAgeRestricted = getBasketResponse(CHECKOUT_METHOD_DELIVERY, 'ageRestriction');

// getAvailableTimes
export const checkoutAvailableFulfilment = getAvailableTimes();

// getSplitNotes
export const splitNotesConfig = getNotesConfig(true);
