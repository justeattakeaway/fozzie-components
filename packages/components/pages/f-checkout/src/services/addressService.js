const storedLocationKey = 'je-full-address-details';

function isFullPostCode (postcode, tenant) {
    const countryPostcodeLengths = {
        uk: postcode.length >= 5,
        au: postcode.length === 4,
        nz: postcode.length === 4
    };

    return !postcode ? false : countryPostcodeLengths[tenant];
}

const formatPostcode = postcode => postcode.replace(/\s/g, '').replace(/-/g, '');

function toFormattedPostcode (postcode, tenant) {
    if (!postcode) {
        return '';
    }

    let formatted = formatPostcode(postcode);

    if (isFullPostCode(formatted, tenant) && tenant === 'uk') {
        const last3 = formatted.slice(formatted.length - 3);
        const prefix = formatted.slice(0, formatted.length - 3);
        formatted = `${prefix} ${last3}`;
    }

    return formatted.toUpperCase();
}

function getAddress (postcode, tenant, address) {
    const shouldShowAdministrativeArea = tenant === 'au';

    if (!address) {
        return {
            line1: '',
            line2: '',
            locality: '',
            ...(shouldShowAdministrativeArea ? { administrativeArea: '' } : {}),
            postcode: toFormattedPostcode(postcode, tenant)
        };
    }

    const lines = [address.Line2, address.Line3];

    return {
        line1: address.Line1,
        line2: lines.filter(l => l).join(', '),
        locality: address.City,
        ...(shouldShowAdministrativeArea ? { administrativeArea: address.Line4 } : {}),
        postcode: address.ZipCode
    };
}

function getDefaultAddress (addressData, tenant) {
    const addresses = addressData.Addresses;

    return !addresses ? null : addresses.find(address => {
        if (tenant === 'uk') {
            return address && address.IsDefault;
        }

        const defaultAddressId = addressData.DefaultAddress;

        return address.AddressId === defaultAddressId;
    });
}

function getAddressClosestToPostcode (postcode, addressData, tenant) {
    if (!postcode) {
        return getDefaultAddress(addressData, tenant);
    }

    const addresses = addressData.Addresses;

    let formattedPostcode = formatPostcode(postcode);

    let address = addresses.find(a => formatPostcode(a.ZipCode) === formattedPostcode);

    if (isFullPostCode(formattedPostcode, tenant)) {
        formattedPostcode = formattedPostcode.slice(0, formattedPostcode.length - 3);
    }



    address = address || addresses.find(a => a && a.ZipCode.startsWith(formattedPostcode));

    return address;
}

/**
 * Checks whether the user has an address in local storage available to us
 * @returns {boolean} - whether the address is in local storage or now
 */
const isAddressInLocalStorage = () => {
    if (window.localStorage) {
        const address = window.localStorage.getItem(storedLocationKey) ? JSON.parse(window.localStorage.getItem(storedLocationKey)) : null;
        return !!(address && (address.PostalCode || address.ZipCode));
    }

    return false;
};

/**
 * Maps the address to how it is returned from the API
 * @param address - The address from local storage
 * @returns The address but mapped to how the API shapes it
 */
function getMappedLocalStorageAddress (address) {
    return {
        lines: [address.Line1, address.Line2, address.Line3],
        locality: address.City,
        postalCode: address.PostalCode || address.ZipCode
    };
}

/**
 * Retrieves the address from local storage
 * @param mapped - boolean deciding whether to return the address in a mapped state
 * @returns address - The address from local storage
 */
function getAddressFromLocalStorage (mapped = true) {
    if (window.localStorage) {
        const storedLocation = window.localStorage.getItem(storedLocationKey);

        if (storedLocation) {
            return mapped ? getMappedLocalStorageAddress(JSON.parse(storedLocation)) : JSON.parse(storedLocation);
        }
    }

    return null;
}

function getAddressCoordsFromLocalStorage () {
    if (window.localStorage) {
        const storedLocation = window.localStorage.getItem(storedLocationKey);
        const parsedLocation = JSON.parse(storedLocation);

        return parsedLocation.Field1 && parsedLocation.Field2 ? [parsedLocation.Field1, parsedLocation.Field2] : null;
    }

    return null;
}

/**
 * Checks whether the address values in local storage match what's is in the form.
 * @param storedAddress - The address stored in local storage
 * @param formAddress - The address in form state
 * @returns {boolean} - Whether the form values match
 */
function doesAddressInStorageAndFormMatch (storedAddress, formAddress) {
    return storedAddress.Line1 === formAddress.line1 && storedAddress.PostalCode === formAddress.postcode
        && storedAddress.Line2 === formAddress.line2 && storedAddress.City === formAddress.locality;
}

/**
 * Attempts to find the closest address; from the addresses
 * supplied, to the current postcode provided.
 * @param addresses - The consumers address book
 * @param tenant - The current tenant
 * @returns {Object} - The closest formatted address
 */
function getClosestAddress (addressData, tenant, currentPostcode) {
    const postcode = currentPostcode || '';
    const address = getAddressClosestToPostcode(postcode, addressData, tenant);

    return getAddress(postcode, tenant, address);
}

export default {
    doesAddressInStorageAndFormMatch,
    getClosestAddress,
    getAddressFromLocalStorage,
    getAddressCoordsFromLocalStorage,
    isAddressInLocalStorage
};
