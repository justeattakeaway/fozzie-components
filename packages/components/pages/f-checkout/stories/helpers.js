import {
    CHECKOUT_NOTE_TYPE_KITCHEN,
    CHECKOUT_NOTE_TYPE_COURIER
} from '../src/constants';

export const NOTE_TYPES = {
    courier: [CHECKOUT_NOTE_TYPE_COURIER],
    split: [CHECKOUT_NOTE_TYPE_COURIER, CHECKOUT_NOTE_TYPE_KITCHEN]
};

export const TENANTS = {
    uk: 'uk',
    au: 'au',
    nz: 'nz'
};

export const DELIVERY_TIMES = {
    later: 'later',
    unavailable: 'unavailable',
    now: 'now'
};

export const httpStatusCodes = {
    noResponse: 0,
    ok: 200,
    badRequest: 400,
    forbidden: 403,
    internalServerError: 500
};

export const httpMethods = {
    get: 'get',
    post: 'post',
    patch: 'patch'
};

export const ISSUES = {
    timeUnavailable: 'FULFILMENT_TIME_UNAVAILABLE',
    serviceTypeUnavailable: 'SERVICE_TYPE_UNAVAILABLE',
    restaurantNotTakingOrders: 'RESTAURANT_NOT_TAKING_ORDERS',
    geoLocationRequired: 'GEOLOCATION_REQUIRED',
    additionalItemsRequired: 'ADDITIONAL_ITEMS_REQUIRED',
    forbidden: 'forbidden'
};
