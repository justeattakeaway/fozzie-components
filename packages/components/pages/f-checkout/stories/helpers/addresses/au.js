const defaultId = 17731860;

const defaultAddress = {
    City: 'Sydney',
    ZipCode: '2089',
    Line1: '196 Kurraba Road',
    Line2: 'Unit 5',
    Line3: 'Neutral Bay',
    Line4: 'New South Wales',
    AddressId: defaultId
};

const additionalAddress = {
    City: 'Melbourne',
    ZipCode: '3071',
    Line1: '94 Shaftesbury Parade',
    Line2: null,
    Line3: 'Thornbury',
    Line4: 'Victoria',
    AddressId: 18086210
};

const defaultGeoLocation = {
    latitude: 27.8606,
    longitude: 152.9788
};

export const auAddresses = { Addresses: [defaultAddress, additionalAddress], DefaultAddress: defaultId };

export const auAddress = {
    address: {
        lines: [defaultAddress.Line1, defaultAddress.Line2, defaultAddress.Line3],
        locality: defaultAddress.City,
        administrativeArea: defaultAddress.Line4,
        postalCode: defaultAddress.ZipCode
    },
    geolocation: defaultGeoLocation
};
