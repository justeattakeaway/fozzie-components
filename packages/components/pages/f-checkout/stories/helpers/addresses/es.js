const defaultId = 17731860;

const defaultAddress = {
    City: 'Barcelona',
    ZipCode: '51238',
    Line1: 'Calle del alba 12',
    Line2: 'Entresuelo B',
    Line3: 'Timbre 354',
    AddressId: defaultId
};

const additionalAddress = {
    City: 'Madrid',
    ZipCode: '28012',
    Line1: 'Calle de las Huertas',
    Line2: '16',
    Line3: '',
    AddressId: 18086210
};

const defaultGeoLocation = {
    latitude: 27.8606,
    longitude: 152.9788
};

export const esAddresses = { Addresses: [defaultAddress, additionalAddress], DefaultAddress: defaultId };

export const esAddress = {
    address: {
        lines: [defaultAddress.Line1, defaultAddress.Line2, defaultAddress.Line3],
        locality: defaultAddress.City,
        postalCode: defaultAddress.ZipCode
    },
    geolocation: defaultGeoLocation
};
