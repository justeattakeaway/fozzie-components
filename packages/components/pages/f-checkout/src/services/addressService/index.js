import uk from './tenants/uk';
import au from './tenants/au';
import nz from './tenants/nz';
import it from './tenants/it';

const tenants = {
    uk,
    au,
    nz,
    it
};

const storedLocationKey = 'je-full-address-details';

function toFormattedPostcode (postcode, tenant) {
    const shouldFormatPostCode = tenant === 'uk';

    if (!postcode) {
        return '';
    }

    if (!shouldFormatPostCode) {
        return postcode;
    }

    return tenants[tenant].toFormattedPostcode(postcode);
}

function formatAddressLines (address) {
    let addressLines;

    if (address.lines) {
        addressLines = {
            line1: address.lines[0],
            ...(address.lines[1] && { line2: address.lines[1] }),
            ...(address.lines[2] && { line3: address.lines[2] }),
            ...(address.lines[3] && { line4: address.lines[3] })
        };
    } else {
        addressLines = {
            line1: address.Line1,
            ...(address.Line2 && { line2: address.Line2 }),
            ...(address.Line3 && { line3: address.Line3 }),
            ...(address.Line4 && { line4: address.Line4 })
        };
    }

    return {
        ...address,
        ...addressLines
    };
}

function getDefaultAddress (addressData, tenant) {
    const addresses = addressData.Addresses;

    return !addresses ? null : addresses.find(address => {
        if (tenant === 'uk') {
            return address && address.IsDefault;
        }

        const defaultAddressId = addressData.DefaultAddress;

        return address.AddressId === defaultAddressId;
    });
}

function getAddressClosestToPostcode (currentPostcode, addressData, tenant) {
    const addresses = addressData.Addresses;

    let postcode = toFormattedPostcode(currentPostcode, tenant);
    let address;

    if (addresses) {
        address = addresses.find(a => {
            const addressPostcode = toFormattedPostcode(a.ZipCode, tenant);

            return addressPostcode === postcode;
        });
    }

    if (tenants[tenant].isFullPostCode(postcode)) {
        postcode = postcode.slice(0, postcode.length - 3);
    }

    return address || addresses?.find(a => a && a.ZipCode.startsWith(postcode));
}

/**
 * Checks whether the user has an address in local storage available to us
 * @returns {boolean} - whether the address is in local storage or now
 */
const isAddressInLocalStorage = () => {
    if (window.localStorage) {
        const address = window.localStorage.getItem(storedLocationKey) ? JSON.parse(window.localStorage.getItem(storedLocationKey)) : null;
        return !!(address && (address.PostalCode || address.ZipCode));
    }

    return false;
};

function getStoredAddress () {
    const storedLocation = window.localStorage.getItem(storedLocationKey);

    return storedLocation ? JSON.parse(storedLocation) : null;
}

/**
 * Retrieves the address from local storage
 * @param mapped - boolean deciding whether to return the address in a mapped state
 * @returns address - The address from local storage
 */
function getAddressFromLocalStorage (tenant, mapped = true) {
    if (window.localStorage) {
        const storedAddress = getStoredAddress();

        if (storedAddress) {
            return mapped ? tenants[tenant].formatAddress(formatAddressLines(storedAddress)) : storedAddress;
        }
    }

    return null;
}

function getAddressCoordsFromLocalStorage () {
    if (window.localStorage) {
        const storedAddress = getStoredAddress();

        return storedAddress?.Field1 && storedAddress?.Field2 ? [storedAddress.Field1, storedAddress.Field2] : null;
    }

    return null;
}

/**
 * Checks whether the address values in local storage match what's is in the form.
 * @param storedAddress - The address stored in local storage
 * @param formAddress - The address in form state
 * @returns {boolean} - Whether the form values match
 */
function doesAddressInStorageAndFormMatch (storedAddress, formAddress) {
    return storedAddress.Line1 === formAddress.line1 && storedAddress.PostalCode === formAddress.postcode
        && storedAddress.Line2 === formAddress.line2 && storedAddress.City === formAddress.locality;
}

/**
 * Attempts to find the closest address; from the addresses
 * supplied, to the current postcode provided.
 * @param addresses - The consumers address book
 * @param tenant - The current tenant
 * @returns {Object} - The closest formatted address
 */
function getClosestAddress (addressData, tenant, currentPostcode) {
    const postcode = currentPostcode || '';
    let address;

    if (!postcode) {
        address = getDefaultAddress(addressData, tenant);
    } else {
        address = getAddressClosestToPostcode(postcode, addressData, tenant);
    }

    if (!address) {
        return tenants[tenant].getEmptyAddress(postcode);
    }

    return tenants[tenant].formatAddress(formatAddressLines(address));
}

export default {
    doesAddressInStorageAndFormMatch,
    getClosestAddress,
    getAddressFromLocalStorage,
    getAddressCoordsFromLocalStorage,
    isAddressInLocalStorage
};
