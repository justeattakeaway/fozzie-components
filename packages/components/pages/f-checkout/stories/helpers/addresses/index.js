import { ukAddress, ukAddresses } from './uk';
import { auAddress, auAddresses } from './au';
import { nzAddress, nzAddresses } from './nz';

export const Address = {
    uk: ukAddress,
    au: auAddress,
    nz: nzAddress
};

export const Addresses = {
    uk: ukAddresses,
    au: auAddresses,
    nz: nzAddresses
};
