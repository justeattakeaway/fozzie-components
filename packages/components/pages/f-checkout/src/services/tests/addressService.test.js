import { when } from 'jest-when';
import addressService from '../addressService';
import localStorageMock from '../../../test-utils/local-storage/local-storage-mock';
import { Addresses } from '../../../stories/helpers/addresses';

const area511Line = {
    City: 'Area 51',
    ZipCode: 'AR51 1AA',
    IsDefault: false,
    Line1: '1 Test Road',
    Line2: null,
    Line3: null
};

const bristol2Lines = {
    City: 'Bristol',
    ZipCode: 'BS1 1AA',
    IsDefault: false,
    Line1: '1 Bristol Road',
    Line2: 'Flat 1',
    Line3: null
};

const london3LinesDefault = {
    City: 'London',
    ZipCode: 'EC4M 7RF',
    IsDefault: true,
    Line1: 'Fleet Place House',
    Line2: 'Farringdon',
    Line3: 'City of London'
};

const defaultAustraliaId = 17731860;

const australiaAddressDefault = {
    City: 'Sydney',
    ZipCode: '2089',
    Line1: '196 Kurraba Road',
    Line2: 'Unit 5',
    Line3: 'Neutral Bay',
    Line4: 'New South Wales',
    AddressId: defaultAustraliaId
};

const australiaAddressAdditional = {
    City: 'Melbourne',
    ZipCode: '3071',
    Line1: '94 Shaftesbury Parade',
    Line2: null,
    Line3: 'Thornbury',
    Line4: 'Victoria',
    AddressId: 18086210
};

describe('addressService', () => {
    describe('getClosestAddress ::', () => {
        const ukAddressesData = { Addresses: [area511Line, bristol2Lines, london3LinesDefault] };
        const auAddressesData = { Addresses: [australiaAddressDefault, australiaAddressAdditional], DefaultAddress: defaultAustraliaId };
        const itAddressesData = Addresses.it;

        describe.each([
            ['uk', ukAddressesData],
            ['au', auAddressesData],
            ['it', itAddressesData]
        ])('when tenant === `%s` AND last searched postcode is not present', (tenant, addressData) => {
            const postcode = '';

            it('should return empty address if addresses is empty', () => {
                // Act
                const actual = addressService.getClosestAddress({}, tenant, postcode);

                // Assert
                expect(actual).toMatchSnapshot();
            });

            it('should return default address if default address is present', () => {
                // Act
                const actual = addressService.getClosestAddress(addressData, tenant, postcode);

                // Assert
                expect(actual).toMatchSnapshot();
            });
        });

        describe.each([
            ['uk', ukAddressesData],
            ['au', auAddressesData],
            ['it', itAddressesData]
        ])('when tenant === `%s` AND last searched postcode is present', (tenant, addressData) => {
            const postcodeTypes = {
                noMatch: {
                    au: '4278',
                    it: '35674',
                    uk: 'EN1 1AA'
                },
                noMatchOrSpaces: {
                    au: '4278',
                    it: '35674',
                    uk: 'EN11AA'
                },
                match: {
                    au: '2089',
                    it: '26844',
                    uk: 'BS1 1AA'
                },
                matchNoSpaces: {
                    au: '2089',
                    it: '26844',
                    uk: 'EC4M7RF'
                },
                partial: {
                    au: '30',
                    it: '20',
                    uk: 'AR51'
                }
            };

            it('should return empty address with postcode set, when postcode does not match', () => {
                // Act
                const actual = addressService.getClosestAddress(addressData, tenant, postcodeTypes.noMatch[tenant]);

                // Assert
                expect(actual).toMatchSnapshot();
            });

            it('should return empty address with formatted postcode set, when postcode does not match and has no spaces', () => {
                // Act
                const actual = addressService.getClosestAddress(addressData, tenant, postcodeTypes.noMatchOrSpaces[tenant]);

                // Assert
                expect(actual).toMatchSnapshot();
            });

            it('should return address with full postcode match', () => {
                // Act
                const actual = addressService.getClosestAddress(addressData, tenant, postcodeTypes.match[tenant]);

                // Assert
                expect(actual).toMatchSnapshot();
            });

            it('should return address with full postcode match when address has no spaces', () => {
                // Act
                const actual = addressService.getClosestAddress(addressData, tenant, postcodeTypes.matchNoSpaces[tenant]);

                // Assert
                expect(actual).toMatchSnapshot();
            });

            it('should return address with postcode partial match', () => {
                // Act
                const actual = addressService.getClosestAddress(addressData, tenant, postcodeTypes.partial[tenant]);

                // Assert
                expect(actual).toMatchSnapshot();
            });
        });
    });

    describe('isAddressInLocalStorage ::', () => {
        describe('if localStorage exists', () => {
            beforeEach(() => {
                Object.defineProperty(window, 'localStorage', { value: localStorageMock });
            });

            afterEach(() => {
                window.localStorage.clear();
                jest.resetAllMocks();
            });

            describe('when the address does NOT exist in local storage', () => {
                it('should return false', () => {
                    expect(addressService.isAddressInLocalStorage()).toBe(false);
                });
            });

            describe('when the address does exist in local storage', () => {
                it('should return true if there is a postcode', async () => {
                    // Arrange
                    window.localStorage.setItem('je-full-address-details', JSON.stringify(london3LinesDefault));

                    // Act & Assert
                    expect(addressService.isAddressInLocalStorage()).toBe(true);
                });

                it('should return false if the customer only searched a postcode', async () => {
                    // Arrange
                    window.localStorage.setItem('je-full-address-details', JSON.stringify({
                        City: null,
                        Field1: null,
                        Field2: null,
                        Line1: null,
                        Line2: null,
                        Line3: null,
                        Line4: null,
                        Line5: null,
                        PostalCode: null,
                        searchBoxAddress: 'BS3 4RL'
                    }));

                    // Act & Assert
                    expect(addressService.isAddressInLocalStorage()).toBe(false);
                });
            });
        });

        describe('if localStorage does NOT exist', () => {
            beforeAll(() => {
                Object.defineProperty(window, 'localStorage', { value: null });
            });

            afterAll(() => {
                window.localStorage.clear();
                jest.resetAllMocks();
            });

            it('should return false', () => {
                // Assert
                expect(addressService.isAddressInLocalStorage()).toBe(false);
            });
        });
    });

    describe('getAddressFromLocalStorage ::', () => {
        beforeEach(() => {
            Object.defineProperty(window, 'localStorage', { value: localStorageMock });
        });

        afterEach(() => {
            window.localStorage.clear();
            jest.resetAllMocks();
        });

        describe('when the address does NOT exist in local storage', () => {
            it('should return null', () => {
                // Assert
                expect(addressService.getAddressFromLocalStorage()).toBe(null);
            });
        });

        describe('when the address does exist in local storage', () => {
            it('should return the address mapped correctly', () => {
                // Arrange
                const expectedAddress = {
                    line1: 'Fleet Place House',
                    line2: 'Farringdon',
                    locality: 'London',
                    postcode: 'EC4M 7RF'
                };
                window.localStorage.setItem('je-full-address-details', JSON.stringify(london3LinesDefault));

                // Act & Assert
                expect(addressService.getAddressFromLocalStorage('uk')).toEqual(expectedAddress);
            });
        });
    });

    describe('getAddressCoords ::', () => {
        it('should return coordinates in correct order', () => {
            // Arrange
            const address = {
                Line1: 'Flat 101',
                Line2: 'Made Up House',
                Line3: 'Camden Town',
                Line4: '',
                Line5: '',
                City: 'London',
                Field1: '51.529747',
                Field2: '-0.142396',
                PostalCode: 'NW1 4DE',
                searchBoxAddress: 'NW1 4DE'
            };

            const expectedCoords = ['51.529747', '-0.142396'];

            // Act & Assert
            expect(addressService.getAddressCoords(address)).toEqual(expectedCoords);
        });

        it('should return coordinates in reverse order if shouldReverse is truthy', () => {
            // Arrange
            const address = {
                Line1: 'Flat 101',
                Line2: 'Made Up House',
                Line3: 'Camden Town',
                Line4: '',
                Line5: '',
                City: 'London',
                Field1: '51.529747',
                Field2: '-0.142396',
                PostalCode: 'NW1 4DE',
                searchBoxAddress: 'NW1 4DE'
            };

            const expectedCoords = ['-0.142396', '51.529747'];

            // Act & Assert
            expect(addressService.getAddressCoords(address, true)).toEqual(expectedCoords);
        });

        it('should return null if coordinates are missing in address', () => {
            // Arrange
            const address = {
                Line1: 'Flat 101',
                Line2: 'Made Up House',
                Line3: 'Camden Town',
                Line4: '',
                Line5: '',
                City: 'London',
                Field1: null,
                Field2: null,
                PostalCode: 'NW1 4DE',
                searchBoxAddress: 'NW1 4DE'
            };

            // Act & Assert
            expect(addressService.getAddressCoords(address)).toEqual(null);
        });
    });

    describe('getAddressFromCookie ::', () => {
        const tenant = 'au';
        it('when the input cookies parameter is invalid should return null', () => {
            // Arrange
            const invalidCookies = {};
            // Act & Assert
            expect(addressService.getAddressFromCookie(tenant, invalidCookies)).toBe(null);
        });

        it('when the address does NOT exists in cookies returns null', () => {
            // Arrange
            const emptyCookies = { get: jest.fn().mockReturnValue(null) };
            // Act & Assert
            expect(addressService.getAddressFromCookie(tenant, emptyCookies)).toBe(null);
        });

        it('when the address in cookies is incomplete returns null', () => {
            // Arrange
            const get = jest.fn();
            when(get).calledWith('je-last_houseNo_used').mockReturnValue('1');
            const incompleteCookies = { get };

            // Act & Assert
            expect(addressService.getAddressFromCookie(tenant, incompleteCookies)).toBe(null);
        });

        describe('when the address does exist in cookies', () => {
            let get;
            let cookies;
            const houseNumber = '35';
            const street = 'Myrtle Bank Road';
            const unitNumber = 'Flat 1';
            const latitude = '54.24';
            const longitude = '0.24';
            const sublocality = '1';
            const state = 'TAS';
            const postalCode = '7259';
            const city = 'Myrtle Bank';
            const searchBoxAddress = '35 Myrtle Bank Road Flat 1, Myrtle Bank 7259 TAS, Australia';

            beforeEach(() => {
                // Arrange - cookies
                get = jest.fn();
                when(get).calledWith('je-last_houseNo_used').mockReturnValue(houseNumber);
                when(get).calledWith('je-last_street_used').mockReturnValue(street);
                when(get).calledWith('je-last_unitNumber_used').mockReturnValue(unitNumber);
                when(get).calledWith('je-last_latitude_used').mockReturnValue(latitude);
                when(get).calledWith('je-last_longitude_used').mockReturnValue(longitude);
                when(get).calledWith('je-last_sublocality_used').mockReturnValue(sublocality);
                when(get).calledWith('je-last_state_used').mockReturnValue(state);
                when(get).calledWith('je-location').mockReturnValue(postalCode);
                when(get).calledWith('je-last_city_used').mockReturnValue(city);
                when(get).calledWith('je-last_suggestion_used').mockReturnValue(searchBoxAddress);
                cookies = { get };
            });

            it('return correct address', () => {
                // Arrange
                const expectedAddress = {
                    line1: `${houseNumber} ${street}`,
                    line2: unitNumber,
                    locality: city,
                    administrativeArea: state,
                    postcode: postalCode
                };

                // Act & Assert
                expect(addressService.getAddressFromCookie(tenant, cookies)).toStrictEqual(expectedAddress);
            });

            it('return correct unformatted address when mapped is falsy', () => {
                // Arrange
                const expectedAddress = {
                    Line1: `${houseNumber} ${street}`,
                    Line2: unitNumber,
                    Line3: sublocality,
                    administrativeArea: state,
                    City: city,
                    PostalCode: postalCode,
                    Field1: latitude,
                    Field2: longitude,
                    searchBoxAddress
                };
                // Act & Assert
                expect(addressService.getAddressFromCookie(tenant, cookies, false)).toStrictEqual(expectedAddress);
            });
        });
    });

    describe('doesAddressInStorageAndFormMatch ::', () => {
        describe('when the address fields are the same', () => {
            it('should return true', () => {
                // Arrange
                const storageAddress = {
                    Line1: 'Fleet Place House',
                    Line2: 'Farringdon',
                    City: 'London',
                    PostalCode: 'EC4M 7RF'
                };

                const formAddress = {
                    line1: 'Fleet Place House',
                    line2: 'Farringdon',
                    locality: 'London',
                    postcode: 'EC4M 7RF'
                };

                // Act & Assert
                expect(addressService.doesAddressInStorageAndFormMatch(storageAddress, formAddress)).toEqual(true);
            });
        });

        describe('when the city field is different', () => {
            it('should return false', () => {
                // Arrange
                const storageAddress = {
                    Line1: 'Fleet Place House',
                    Line2: 'Farringdon',
                    City: 'Newcastle',
                    PostalCode: 'EC4M 7RF'
                };

                const formAddress = {
                    line1: 'Fleet Place House',
                    line2: 'Farringdon',
                    locality: 'London',
                    postcode: 'EC4M 7RF'
                };

                // Act & Assert
                expect(addressService.doesAddressInStorageAndFormMatch(storageAddress, formAddress)).toEqual(false);
            });
        });

        describe('when the postcode is different', () => {
            it('should return false', () => {
                // Arrange
                const storageAddress = {
                    Line1: 'Fleet Place House',
                    Line2: 'Farringdon',
                    City: 'London',
                    PostalCode: 'EC4M 7RF'
                };

                const formAddress = {
                    line1: 'Fleet Place House',
                    line2: 'Farringdon',
                    locality: 'London',
                    postcode: 'EC4M 7RE'
                };

                // Act & Assert
                expect(addressService.doesAddressInStorageAndFormMatch(storageAddress, formAddress)).toEqual(false);
            });
        });
    });

    describe('setAddressInLocalStorage ::', () => {
        const addressDetails = {
            postcode: 'TA3 1SS',
            line1: '86 Pier Road',
            line2: 'Flat 1',
            line3: null,
            line4: null,
            administrativeArea: null,
            locality: 'Stapley'
        };

        describe('when local storage is available', () => {
            it('save address in local storage correctly', () => {
                // Arrange
                const spy = jest.spyOn(window.localStorage, 'setItem');
                const expectedValue = JSON.stringify({
                    PostalCode: addressDetails.postcode,
                    Line1: addressDetails.line1,
                    Line2: addressDetails.line2,
                    Line3: addressDetails.line3,
                    Line4: addressDetails.line4,
                    administrativeArea: addressDetails.administrativeArea,
                    City: addressDetails.locality
                });

                // Act
                addressService.setAddressInLocalStorage(addressDetails);

                // Assert
                expect(spy).toHaveBeenCalledWith('je-full-address-details', expectedValue);
            });
        });
    });
});
