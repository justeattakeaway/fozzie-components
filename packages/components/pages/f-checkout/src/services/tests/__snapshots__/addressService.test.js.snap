// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is not present should return default address if default address is present 1`] = `
Object {
  "administrativeArea": "New South Wales",
  "line1": "196 Kurraba Road",
  "line2": "Unit 5, Neutral Bay",
  "locality": "Sydney",
  "postcode": "2089",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is not present should return empty address if addresses is empty 1`] = `
Object {
  "administrativeArea": "",
  "line1": "",
  "line2": "",
  "locality": "",
  "postcode": "",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is present should return address with full postcode match 1`] = `
Object {
  "administrativeArea": "New South Wales",
  "line1": "196 Kurraba Road",
  "line2": "Unit 5, Neutral Bay",
  "locality": "Sydney",
  "postcode": "2089",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is present should return address with full postcode match when address has no spaces 1`] = `
Object {
  "administrativeArea": "New South Wales",
  "line1": "196 Kurraba Road",
  "line2": "Unit 5, Neutral Bay",
  "locality": "Sydney",
  "postcode": "2089",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is present should return address with postcode partial match 1`] = `
Object {
  "administrativeArea": "Victoria",
  "line1": "94 Shaftesbury Parade",
  "line2": "Thornbury",
  "locality": "Melbourne",
  "postcode": "3071",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is present should return empty address with formatted postcode set, when postcode does not match and has no spaces 1`] = `
Object {
  "administrativeArea": "",
  "line1": "",
  "line2": "",
  "locality": "",
  "postcode": "4278",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`au\` AND last searched postcode is present should return empty address with postcode set, when postcode does not match 1`] = `
Object {
  "administrativeArea": "",
  "line1": "",
  "line2": "",
  "locality": "",
  "postcode": "4278",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is not present should return default address if default address is present 1`] = `
Object {
  "line1": "Fleet Place House",
  "line2": "Farringdon, City of London",
  "locality": "London",
  "postcode": "EC4M 7RF",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is not present should return empty address if addresses is empty 1`] = `
Object {
  "line1": "",
  "line2": "",
  "locality": "",
  "postcode": "",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is present should return address with full postcode match 1`] = `
Object {
  "line1": "1 Bristol Road",
  "line2": "Flat 1",
  "locality": "Bristol",
  "postcode": "BS1 1AA",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is present should return address with full postcode match when address has no spaces 1`] = `
Object {
  "line1": "Fleet Place House",
  "line2": "Farringdon, City of London",
  "locality": "London",
  "postcode": "EC4M 7RF",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is present should return address with postcode partial match 1`] = `
Object {
  "line1": "1 Test Road",
  "line2": "",
  "locality": "Area 51",
  "postcode": "AR51 1AA",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is present should return empty address with formatted postcode set, when postcode does not match and has no spaces 1`] = `
Object {
  "line1": "",
  "line2": "",
  "locality": "",
  "postcode": "EN1 1AA",
}
`;

exports[`addressService getClosestAddress :: when tenant === \`uk\` AND last searched postcode is present should return empty address with postcode set, when postcode does not match 1`] = `
Object {
  "line1": "",
  "line2": "",
  "locality": "",
  "postcode": "EN1 1AA",
}
`;
